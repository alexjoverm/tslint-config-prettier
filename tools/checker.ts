import * as path from "path";
import { Configuration, Linter } from "tslint";

// tslint:disable-next-line:no-var-requires
const tslintConfigPrettier = require("..") as Configuration.RawConfigFile;

export const check = (configFilePaths: string[]) => {
  if (configFilePaths.length === 0) {
    // tslint:disable-next-line:no-console
    console.log("Usage: tslint-config-prettier-check <pathToConfigFile> ...");
    return;
  }

  configFilePaths.forEach((configFilePath) => {
    try {
      const conflictRules = getConflictRules(configFilePath);
      if (conflictRules.length === 0) {
        // tslint:disable-next-line:no-console
        console.log(`No conflict rule detected in ${configFilePath}`);
      } else {
        // tslint:disable-next-line:no-console
        console.error(
          `Conflict rule(s) detected in ${configFilePath}:\n${conflictRules
            .map((conflictRule) => `  ${conflictRule}`)
            .join("\n")}`,
        );
        process.exitCode = 1;
      }
    } catch (error) {
      // tslint:disable-next-line:no-console
      console.error(error.message);
      process.exitCode = 1;
    }
  });
};

function getConflictRules(configFilePath: string) {
  const { rules, jsRules } = Linter.loadConfigurationFromPath(configFilePath);
  return Object.keys(tslintConfigPrettier.rules!).filter(
    (conflictRuleName) =>
      isConflict(conflictRuleName, rules) ||
      isConflict(conflictRuleName, jsRules),
  );
}

function isConflict(
  conflictRuleName: string,
  rules: Configuration.IConfigurationFile["rules"],
) {
  return (
    rules.has(conflictRuleName) &&
    rules.get(conflictRuleName)!.ruleSeverity !== "off"
  );
}
