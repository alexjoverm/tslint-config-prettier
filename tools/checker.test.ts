import * as path from "path";
import { check } from "./checker";

expect.addSnapshotSerializer({
  print: (value: string, serializer) =>
    serializer(value.replace(process.cwd(), "<cwd>")),
  test: (value) =>
    typeof value === "string" && value.indexOf(process.cwd()) !== -1,
});

const messages: string[] = [];

mockConsole("log");
mockConsole("error");

beforeEach(() => {
  messages.splice(0, messages.length);
  process.exitCode = 0;
});

afterAll(() => {
  process.exitCode = 0;
});

it("should throw conflict rules", () => {
  checkFixture("all");
  expect(getMessage()).toMatchSnapshot();
  expect(process.exitCode).toBe(1);
});

it("should throw parsing error for invalid config", () => {
  checkFixture("error");
  expect(getMessage()).toMatchSnapshot();
  expect(process.exitCode).toBe(1);
});

it("should show success message for non-conflict rules", () => {
  checkFixture("empty");
  expect(getMessage()).toMatchSnapshot();
  expect(process.exitCode).toBe(0);
});

it("should show success message for conflict rules that is disabled", () => {
  checkFixture("false");
  expect(getMessage()).toMatchSnapshot();
  expect(process.exitCode).toBe(0);
});

it("should show usage if there is no filePath", () => {
  checkFixture();
  expect(getMessage()).toMatchSnapshot();
  expect(process.exitCode).toBe(0);
});

it("should be able to check multiple config files", () => {
  checkFixture("all", "empty", "error");
  expect(getMessage()).toMatchSnapshot();
  expect(process.exitCode).toBe(1);
});

function mockConsole(id: keyof typeof console) {
  const mockedConsoleLog = jest
  .spyOn(console, id)
  .mockImplementation((message: string) =>
    messages.push(
      message
        .split("\n")
        .map((x) => `[${id}] ${x}`)
        .join("\n"),
    ),
  );
}

function getMessage() {
  return messages.join("\n");
}

function checkFixture(...ids: string[]) {
  const filePaths = ids.map((id) =>
    path.resolve(__dirname, `../fixtures/tslint.${id}.json`),
  );
  return check(filePaths);
}
