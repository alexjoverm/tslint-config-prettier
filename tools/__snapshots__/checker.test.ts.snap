// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to check multiple config files 1`] = `
"[error] Conflict rule(s) detected in <cwd>/fixtures/tslint.all.json:
[error]   align
[error]   arrow-parens
[error]   eofline
[error]   import-spacing
[error]   indent
[error]   linebreak-style
[error]   max-line-length
[error]   new-parens
[error]   newline-per-chained-call
[error]   no-consecutive-blank-lines
[error]   no-irregular-whitespace
[error]   no-trailing-whitespace
[error]   number-literal-format
[error]   object-literal-key-quotes
[error]   one-line
[error]   quotemark
[error]   semicolon
[error]   space-before-function-paren
[error]   space-within-parens
[error]   trailing-comma
[error]   typedef-whitespace
[error]   whitespace
[error] Unexpected number in JSON at position 1 in <cwd>/fixtures/tslint.error.json"
`;

exports[`should show nothing for conflict rules that is disabled 1`] = `""`;

exports[`should show nothing for non-conflict rules 1`] = `""`;

exports[`should show usage if there is no filePath 1`] = `"[log] Usage: tslint-config-prettier-check <pathToConfigFile> ..."`;

exports[`should throw conflict rules 1`] = `
"[error] Conflict rule(s) detected in <cwd>/fixtures/tslint.all.json:
[error]   align
[error]   arrow-parens
[error]   eofline
[error]   import-spacing
[error]   indent
[error]   linebreak-style
[error]   max-line-length
[error]   new-parens
[error]   newline-per-chained-call
[error]   no-consecutive-blank-lines
[error]   no-irregular-whitespace
[error]   no-trailing-whitespace
[error]   number-literal-format
[error]   object-literal-key-quotes
[error]   one-line
[error]   quotemark
[error]   semicolon
[error]   space-before-function-paren
[error]   space-within-parens
[error]   trailing-comma
[error]   typedef-whitespace
[error]   whitespace"
`;

exports[`should throw parsing error for invalid config 1`] = `"[error] Unexpected number in JSON at position 1 in <cwd>/fixtures/tslint.error.json"`;
